lib/metasploit/framework/login_scanner/snmp.rb

```diff
@@ -54,93 +60,286 @@ def versions
+              send_snmp_write_request(version, community, data)
             end
-          rescue RuntimeError
-            return false
+            sleep_between_attempts
           end
+          process_responses(1.0)
+        end
+
+        # Process any responses on the UDP socket and queue the results
+        def process_responses(timeout=1.0)
+          queue = []
+          while (res = sock.recvfrom(65535, timeout))
+
+            # Ignore invalid responses
+            break if not res[1]
+
+            # Ignore empty responses
+            next if not (res[0] and res[0].length > 0)
+
+            # Trim the IPv6-compat prefix off if needed
+            shost = res[1].sub(/^::ffff:/, '')
 
+            response = parse_snmp_response(res[0])
+            next unless response
+
+            self.queued_results << {
+              community: response[:community],
+              host: host,
+              port: port,
+              protocol: 'udp',
+              service_name: 'snmp',
+              proof: response[:proof],
+              status: Metasploit::Model::Login::Status::SUCCESSFUL,
+              access_level: 'read-only',
+              snmp_version: response[:version],
+              snmp_error: response[:error]
+            }
+          end
         end
 
-        # Sets the connection timeout appropriately for SNMP
-        # if the user did not set it.
-        def set_sane_defaults
-          self.connection_timeout = DEFAULT_TIMEOUT if self.connection_timeout.nil?
-          self.port = DEFAULT_PORT if self.port.nil?
-          self.retries = DEFAULT_RETRIES if self.retries.nil?
-          self.version = DEFAULT_VERSION if self.version.nil?
+        # Create and send a SNMP read request for sys.sysDescr.0
+        def send_snmp_read_request(version, community)
+          send_snmp_request(
+            create_snmp_read_sys_descr_request(version, community)
+          )
+        end
+
+        # Create and send a SNMP write request for sys.sysDescr.0
+        def send_snmp_write_request(version, community, data)
+          send_snmp_request(
+            create_snmp_write_sys_descr_request(version, community, data)
+          )
         end
 
-        # This method takes an snmp client and tests whether
-        # it has read access to the remote system. It checks
-        # the sysDescr oid to use as proof
-        # @param snmp_client [SNMP::Manager] The SNMP client to use
-        # @return [String, nil] Returns a string if successful, nil if failed
-        def test_read_access(snmp_client)
-          proof = nil
+        # Send a SNMP request on the existing socket
+        def send_snmp_request(pkt)
+          resend_count = 0
+
           begin
-            resp = snmp_client.get("sysDescr.0")
-            resp.each_varbind { |var| proof = var.value }
-          rescue RuntimeError
-            proof = nil
+            sock.sendto(pkt, self.host, self.port, 0)
+          rescue ::Errno::ENOBUFS
```

Comment:
BSD and OSX will raise ` ::Errno::EISCONN` here. I don't really remember why, but the fix I went with when running into a similar problem in the manager class is here:

https://github.com/rapid7/metasploit-framework/pull/3935/files